File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        maxLen = 0
        left = 0
        charFreqMap = {}
        for i in range(len(s)):
            if s[i] in charFreqMap:
                #prevent left pointer from going backwards
                left = max(left,charFreqMap[s[i]] + 1)
                charFreqMap[s[i]] = i
            else:
                charFreqMap[s[i]] = i
            maxLen = max(maxLen, i - left + 1)
        return maxLen

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.go
func lengthOfLongestSubstring(s string) int {
    maxLen := 0
    left := 0
    charFreqMap := make(map[byte]int)
    for i := range(len(s)){
        value,ok := charFreqMap[s[i]]
        if (ok) {
            //prevent left pointer from going backwards
            left = max(left,value+1)
        }
        charFreqMap[s[i]] = i
        maxLen = max(maxLen, i - left + 1)
    }
    return maxLen

}